File: imperative/test_gen.py
'''
Run this file for testing.
'''
import sys
from input.scanner import Scanner
from input.parser import Parser
from input.instruction_buffer import InstructionBuffer
from intermediate.liveness import Liveness
from intermediate.interference_graph import InterferenceGraph
from generator.asm_generator import ASMGenerator

TEST_DIR = "tests/test-output/"
REG_COUNT = 8
# Helpers
def get_generator(input_file: str, num_regs: int) -> ASMGenerator:
    file = open_file(input_file)
    parser = Parser(Scanner(file))
    
    try:
        instruction_buffer = parser.parse()
    except ValueError as ve:
        print(f"Error parsing input: {ve}")
        sys.exit(1)
    
    liveness = Liveness(instruction_buffer)
    variables = instruction_buffer.get_occurred_variables()
    interference_graph = InterferenceGraph()
    interference_graph.build_graph(liveness, variables)
    print(liveness.liveness_info())
    
    try:
        interference_graph.color_graph(num_regs)
    except ValueError as ve:
        print(f"Error colouring graph: {ve}")
        sys.exit(1)
    
    interference_graph.print_variable_interference_table()
    interference_graph.print_register_colouring_table(num_regs)

    return ASMGenerator(instruction_buffer, interference_graph, liveness)

def open_file(input_file: str):
    try:
        file = open("test-input/" + input_file, 'r')
    except FileNotFoundError:
        print(f"File not found: {input_file}")
        sys.exit(1)
    return file

def test_assignments():
    generator: ASMGenerator = get_generator("assignments.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "assignments.s")

def test_binary_ops_with_literals():
    generator: ASMGenerator = get_generator("binary_ops_with_literals.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "binary_ops_with_literals.s")

def test_binary_ops():
    generator: ASMGenerator = get_generator("binary_ops.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "binary_ops.s")

def test_chained_reuse():
    generator: ASMGenerator = get_generator("chained_reuse.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "chained_reuse.s")

def test_empty_file():
    generator: ASMGenerator = get_generator("empty_file.txt", REG_COUNT)    
    generator.generate_assembly(TEST_DIR + "empty_file.s")

def test_invalid_operators():
    generator: ASMGenerator = get_generator("invalid_operators.txt", REG_COUNT)    
    generator.generate_assembly(TEST_DIR + "invalid_operators.s")

def test_malformed_in_middle():
    generator: ASMGenerator = get_generator("malformed_in_middle.txt", REG_COUNT)    
    generator.generate_assembly(TEST_DIR + "malformed_in_middle.s")

def test_malformed_syntax():
    generator: ASMGenerator = get_generator("malformed_syntax.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "malformed_syntax.s")

def test_missing_operands():
    generator: ASMGenerator = get_generator("missing_operands.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "missing_operands.s")

def test_mixed_all_types():
    generator: ASMGenerator = get_generator("mixed_all_types.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "mixed_all_types.s")

def test_multiple_live_vars():
    generator: ASMGenerator = get_generator("multiple_live_vars.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "multiple_live_vars.s")

def test_no_live_declaration():
    generator: ASMGenerator = get_generator("no_live_declaration.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "no_live_declaration.s")

def test_reference_program():
    generator: ASMGenerator = get_generator("reference_program.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "reference_program.s")

def test_single_assignment():
    generator: ASMGenerator = get_generator("single_assignment.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "single_assignment.s")

def test_single_binary():
    generator: ASMGenerator = get_generator("single_binary.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "single_binary.s")

def test_single_unary():
    generator: ASMGenerator = get_generator("single_unary.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "single_unary.s")

def test_unary_ops():
    generator: ASMGenerator = get_generator("unary_ops.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "unary_ops.s")

def test_undefined_variables():
    generator: ASMGenerator = get_generator("undefined_variables.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "undefined_variables.s")

def test_valid_then_invalid_operator():
    generator: ASMGenerator = get_generator("valid_then_invalid_operator.txt", REG_COUNT)
    generator.generate_assembly(TEST_DIR + "valid_then_invalid_operator.s")

if __name__ == "__main__":
    test_assignments() # Want successful ASM output - Success
    test_binary_ops_with_literals() # Want successful ASM output - Success
    test_binary_ops() # Want successful ASM output - Success
    test_chained_reuse() # Want successful ASM output - Success
    test_empty_file() # Want empty '.s' file - Success
    
    try:
        test_invalid_operators() # Looking for error - Error confirmed
    except:
        print("Error successfully")
    
    try:
        test_malformed_syntax() # Looking for error - Error confirmed    
    except:
        print("Error successfully")
    
    try:
        test_missing_operands() # Looking for error - Error confirmed 
    except ValueError as ve:
        print("Error successfully")

    test_mixed_all_types() # Want successful ASM output - Success
    test_multiple_live_vars() # Want successful ASM output - Success
    test_no_live_declaration() # Want ASM output with no MOV back into memory at the end of the output - Success
    test_reference_program() # Want something similar to D2L example - Success. 4 registers are used. D2L example uses 7, optimized D2L version uses 2
    test_single_assignment() # Looking for assignment and then a move back into memory - Success
    test_single_binary() # Looking for MOV, ADD, then MOV back to save program state - Success
    test_single_unary() # Looking for MOV, MUL #-1, MOV - Success
    test_unary_ops() # Looking for multiple instances similar to last test - Success
    test_undefined_variables() # Looking for two MOV into separate registers, add those registers, then MOV back to memory - Success
    try:
        test_valid_then_invalid_operator() # Looking for an error to be raised in the console - Success
    except:
        print("Error successfully")

File: haskell/TestGen.hs
import Control.Exception (SomeException, try)
import Gen (gen)

regCount :: Int
regCount = 8

testDir :: String
testDir = "test-input/"

runTest :: String -> IO ()
runTest inputFile = gen regCount (testDir ++ inputFile)

runTestExpectError :: String -> IO ()
runTestExpectError inputFile = do
  result <- try (runTest inputFile) :: IO (Either SomeException ())
  case result of
    Left ex -> putStrLn $ "Error successfully: " ++ show ex
    Right _  -> putStrLn $ "WARNING: Expected error but none was raised for " ++ inputFile

-- Tests expected to succeed and produce ASM output
testAssignments :: IO ()
testAssignments = runTest "assignments.txt"

testBinaryOpsWithLiterals :: IO ()
testBinaryOpsWithLiterals = runTest "binary_ops_with_literals.txt"

testBinaryOps :: IO ()
testBinaryOps = runTest "binary_ops.txt"

testChainedReuse :: IO ()
testChainedReuse = runTest "chained_reuse.txt"

testEmptyFile :: IO ()
testEmptyFile = runTest "empty_file.txt"

testMixedAllTypes :: IO ()
testMixedAllTypes = runTest "mixed_all_types.txt"

testMixedMany :: IO ()
testMixedMany = runTest "mixed_many.txt"

testMultipleLiveVars :: IO ()
testMultipleLiveVars = runTest "multiple_live_vars.txt"

testNoLiveDeclaration :: IO ()
testNoLiveDeclaration = runTest "no_live_declaration.txt"

testReferenceProgram :: IO ()
testReferenceProgram = runTest "reference_program.txt"

testSingleAssignment :: IO ()
testSingleAssignment = runTest "single_assignment.txt"

testSingleBinary :: IO ()
testSingleBinary = runTest "single_binary.txt"

testSingleUnary :: IO ()
testSingleUnary = runTest "single_unary.txt"

testUnaryOps :: IO ()
testUnaryOps = runTest "unary_ops.txt"

testUndefinedVariables :: IO ()
testUndefinedVariables = runTest "undefined_variables.txt"

-- Tests expected to raise errors
testInvalidOperators :: IO ()
testInvalidOperators = runTestExpectError "invalid_operators.txt"

testMalformedInMiddle :: IO ()
testMalformedInMiddle = runTestExpectError "malformed_in_middle.txt"

testMalformedSyntax :: IO ()
testMalformedSyntax = runTestExpectError "malformed_syntax.txt"

testMissingOperands :: IO ()
testMissingOperands = runTestExpectError "missing_operands.txt"

testValidThenInvalidOperator :: IO ()
testValidThenInvalidOperator = runTestExpectError "valid_then_invalid_operator.txt"

main :: IO ()
main = do
  putStrLn "=== Running TestGen ==="

  putStrLn "\n-- assignments (expect: ASM output) --"
  testAssignments

  putStrLn "\n-- binary_ops_with_literals (expect: ASM output) --"
  testBinaryOpsWithLiterals

  putStrLn "\n-- binary_ops (expect: ASM output) --"
  testBinaryOps

  putStrLn "\n-- chained_reuse (expect: ASM output) --"
  testChainedReuse

  putStrLn "\n-- empty_file (expect: empty .s file) --"
  testEmptyFile

  putStrLn "\n-- invalid_operators (expect: error) --"
  testInvalidOperators

  putStrLn "\n-- malformed_in_middle (expect: error) --"
  testMalformedInMiddle

  putStrLn "\n-- malformed_syntax (expect: error) --"
  testMalformedSyntax

  putStrLn "\n-- missing_operands (expect: error) --"
  testMissingOperands

  putStrLn "\n-- mixed_all_types (expect: ASM output) --"
  testMixedAllTypes

  putStrLn "\n-- mixed_many (expect: ASM output) --"
  testMixedMany

  putStrLn "\n-- multiple_live_vars (expect: ASM output) --"
  testMultipleLiveVars

  putStrLn "\n-- no_live_declaration (expect: ASM output, no MOV back to memory) --"
  testNoLiveDeclaration

  putStrLn "\n-- reference_program (expect: ASM output) --"
  testReferenceProgram

  putStrLn "\n-- single_assignment (expect: assignment then MOV to memory) --"
  testSingleAssignment

  putStrLn "\n-- single_binary (expect: MOV, ADD, MOV to memory) --"
  testSingleBinary

  putStrLn "\n-- single_unary (expect: MOV, MUL #-1, MOV) --"
  testSingleUnary

  putStrLn "\n-- unary_ops (expect: ASM output) --"
  testUnaryOps

  putStrLn "\n-- undefined_variables (expect: ASM output) --"
  testUndefinedVariables

  putStrLn "\n-- valid_then_invalid_operator (expect: error) --"
  testValidThenInvalidOperator

  putStrLn "\n=== TestGen complete ==="
